### ISSUE TITLE
deploy amber contracts to foucoco and test that it works

### ISSUE THREAD

**Issue by ashneverdawn** [pendulum-chain/pendulum#225]
deploy amber contracts to foucoco and test that it works

amber compiled contracts:
[https://app.zenhub.com/files/428984346/5ba8448e-f9df-4eb1-b9ab-76010834a767/download](https://app.zenhub.com/files/428984346/5ba8448e-f9df-4eb1-b9ab-76010834a767/download)

amber compiled slippage curve:
[https://app.zenhub.com/files/428984346/f18f9044-057a-43f3-b8d8-4f88912403d2/download](https://app.zenhub.com/files/428984346/f18f9044-057a-43f3-b8d8-4f88912403d2/download)

amber deployment scripts:
[https://app.zenhub.com/files/428984346/b5e33e2d-68d1-48ee-a79f-b4454e54427d/download](https://app.zenhub.com/files/428984346/b5e33e2d-68d1-48ee-a79f-b4454e54427d/download)

deployment scripts were written for mumbai, so they cannot be used as is, however the deployment steps can be derived from them.

instead of regular erc20 tokens, we should use wrapped native tokens using our psp22pendulum contract found in our `pendulum-ink-wrapper` repo.

**Comment by prayagd** [comment]
@ashneverdawn Can you please update the status here? Looks like the PR is merged.

**Comment by ashneverdawn** [comment]
yes, PR got merged last week. Today, I'll do the runtime upgrade on Foucoco so I can resume the deployment.

**Comment by ashneverdawn** [comment]
### ----- 06_deploy_swap_pools -----

Deploy Contract: SwapPool
Args: [
mUSD.address,
curve.address,
router.address,
backstop.address,
treasury,
"0xAmber mUSD Swap LP",
"mUSD-LP"
]
Contract Address: 6jkq3j4cq119eF6xZDD2NnohtKJnYEhgXpt8rUJgjJSWC5hY

Deploy Contract: SwapPool
Args: [
mEUR.address,
curve.address,
router.address,
backstop.address,
treasury,
"0xAmber mEUR Swap LP",
"mEUR-LP"
]
Contract Address: 6kS2LNht4gpHck8YwPWGxFADP8x1xRcNpYKg1DLBMNXtEeU8

Deploy Contract: SwapPool
Args: [
mGBP.address,
curve.address,
router.address,
backstop.address,
treasury,
"0xAmber mGBP Swap LP",
"mGBP-LP"
]
Contract Address: 6jxtvkaowhQ4RLgegD5wKYDoBspKAcZ7kh7F7ME6nMftXzYr

Call Contract: Router
Method: registerPool
Args: [
mUSD.address,
LPmUSD.address
]
Error:
**Contract Reverted!
DispatchError
ContractTrapped**

**DispatchError docs
Contract trapped during execution.**

**Comment by ashneverdawn** [comment]
I was able to deploy all the swap pools successfully 👍 
However, when trying to call `registerPool` of the `Router` the contract is failing.

This is the registerPool function from the solidity code:
```

    /**
     * @notice Registers a newly created swap pool.
     */
    function registerPool(address _asset, address _swapPool)
        external
        onlyOwner
    {
        require(_asset != address(0), "NO_ASSET");
        require(_swapPool != address(0), "NO_POOL");
        require(address(oracleByAsset[_asset]) != address(0), "NO_ORACLE");

        poolByAsset[_asset] = ISwapPoolPermissioned(_swapPool);
        IERC20(_asset).approve(_swapPool, 2**256  - 1);
    }

```

It's not obvious to me what could be the issue. Will investigate further tomorrow.

**Comment by ashneverdawn** [comment]
Without a way to deploy all the contracts with a script, it's tedious to narrow down this issue. I'm putting this on hold until I've finished creating a way to do that.

**Comment by andywer** [comment]
@ashneverdawn What was the error?

**Comment by ashneverdawn** [comment]
The error was:
```
Contract Reverted!
DispatchError
ContractTrapped

DispatchError docs
Contract trapped during execution.
```

Which isn't very helpful, basically just tells me that the contract threw some kind of exception/panicked

**Comment by ashneverdawn** [comment]
@andywer Once I have a deployment script working, first thing I'll try is deploy the latest version of the contracts. If that doesn't solve it, then I'll dig into this `registerPool` function and check the solang docs: https://solang.readthedocs.io/en/latest/ to see if this is caused by any differences with EVM we may have overlooked.

**Comment by ashneverdawn** [comment]
@andywer yes, I did.

**Comment by ashneverdawn** [comment]
To give an update here: 
- We have resolved the deployment issue a while ago already. 
- We also were able to use the wasm-deploy tool to deploy all the contracts successfully via script. 
- What's missing here is to test that it actually operates as expected. We don't have a good understanding yet on the flow of what needs to be called and in what order. Nabla does have some testing contracts that we could deploy and call with our wasm-deploy tool, however some changes are needed since they rely on some functionality from Foundry which is only for EVM chains.
- Also note we've been semi-blocked by Foucoco being bricked. I say semi-blocked because we can of course still do some things locally. But ultimately we need to make sure this is live and working properly on Foucoco, which is also important for the ui and indexer that depend on this.

@TorstenStueber Since this was mostly worked on by you, is there anything I missed that you'd like to add?
